src/clocksi_static_tx_coord_fsm.erl

```diff
@@ -149,3 +250,87 @@ code_change(_OldVsn, StateName, State, _Extra) -> {ok, StateName, State}.
+%%       starting this transaction has seen, and
+%%     2.machine's local time, as returned by erlang:now().
+-spec get_snapshot_time(DcId :: term(), ClientClock :: vectorclock:vectorclock())
+                       -> {ok, vectorclock:vectorclock()} | {error,term()}.
+get_snapshot_time(DcId, ClientClock) ->
+    wait_for_clock(DcId, ClientClock).
+
+
+perform_operations([], _Transaction, ReadSet, Partitions) ->
+    {ReadSet, Partitions};
+perform_operations([{read, Key, Type}|Rest], Transaction, ReadSet, Partitions) ->
+    Preflist = log_utilities:get_preflist_from_key(Key),
+    IndexNode = hd(Preflist),
+    case clocksi_vnode:read_data_item(IndexNode, Transaction, Key, Type) of
+        {error, Reason} ->
+            {error, Reason};
+        {ok, Snapshot} ->
+            Value = Type:value(Snapshot),
+            perform_operations(Rest, Transaction, ReadSet++[Value], Partitions)
+    end;
+perform_operations([{update, Key, Type, Param}|Rest], Transaction, ReadSet, Partitions) ->
+    Preflist = log_utilities:get_preflist_from_key(Key),
+    IndexNode = hd(Preflist),
+    case generate_downstream_op(Transaction, IndexNode, Key, Type, Param) of
+        {ok, DownstreamRecord} ->
+            case clocksi_vnode:update_data_item(IndexNode, Transaction,
+                Key, Type, DownstreamRecord) of
+                ok ->
+                    case lists:member(IndexNode, Partitions) of
+                        false ->
+                            perform_operations(Rest, Transaction, ReadSet, Partitions++[IndexNode]);
+                        true->
+                            perform_operations(Rest, Transaction, ReadSet, Partitions)
+                    end;
+                {error, Reason} ->
+                    {error, Reason}
+            end;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+
+-spec get_snapshot_time(DcId :: term()) -> {ok, vectorclock:vectorclock()} | {error, term()}.
+get_snapshot_time(DcId) ->
+    Now = clocksi_vnode:now_microsec(erlang:now()),
+    case vectorclock:get_stable_snapshot() of
+        {ok, VecSnapshotTime} ->
+            SnapshotTime = dict:update(DcId,
+                                       fun (_Old) -> Now end,
+                                       Now, VecSnapshotTime),
+            {ok, SnapshotTime};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec wait_for_clock(DcId :: term(), Clock :: vectorclock:vectorclock()) ->
+                           {ok, vectorclock:vectorclock()} | {error, term()}.
+wait_for_clock(DcId, Clock) ->
+   case get_snapshot_time(DcId) of
+       {ok, VecSnapshotTime} ->
+           case vectorclock:ge(VecSnapshotTime, Clock) of
+               true ->
+                   %% No need to wait
+                   {ok, VecSnapshotTime};
+               false ->
+                   %% wait for snapshot time to catch up with Client Clock
+                   timer:sleep(10),
```

Comment:
I guess just a reasonable value for spinning for clock. I copied from interactive_tx_coord